include/deal.II/lac/sparse_matrix_tools.h

```diff
@@ -0,0 +1,592 @@
+                                   SparseMatrixType2 &        system_matrix_out,
+                                   SparsityPatternType2 &sparsity_pattern_out)
+  {
+    restrict_to_serial_sparse_matrix(system_matrix,
+                                     sparsity_pattern,
+                                     requested_is,
+                                     IndexSet(), // simply pass empty index set
+                                     system_matrix_out,
+                                     sparsity_pattern_out);
+  }
+
+
+
+  template <typename SparseMatrixType,
+            typename SparsityPatternType,
+            typename Number>
+  void
+  restrict_to_full_matrices(
+    const SparseMatrixType &                                 system_matrix,
+    const SparsityPatternType &                              sparsity_pattern,
+    const std::vector<std::vector<types::global_dof_index>> &indices,
+    std::vector<FullMatrix<Number>> &                        blocks)
+  {
+    // 0) determine which rows are locally owned and which ones are remote
+    const auto local_size = internal::get_local_size(system_matrix);
+    const auto prefix_sum = internal::compute_prefix_sum(
+      local_size, internal::get_mpi_communicator(system_matrix));
+    IndexSet locally_owned_dofs(std::get<1>(prefix_sum));
+    locally_owned_dofs.add_range(std::get<0>(prefix_sum),
+                                 std::get<0>(prefix_sum) + local_size);
+
+    std::vector<dealii::types::global_dof_index> ghost_indices_vector;
+
+    for (const auto &i : indices)
+      ghost_indices_vector.insert(ghost_indices_vector.end(),
+                                  i.begin(),
+                                  i.end());
+
+    std::sort(ghost_indices_vector.begin(), ghost_indices_vector.end());
+    ghost_indices_vector.erase(std::unique(ghost_indices_vector.begin(),
+                                           ghost_indices_vector.end()),
+                               ghost_indices_vector.end());
+
+
+    IndexSet locally_active_dofs(std::get<1>(prefix_sum));
+    locally_active_dofs.add_indices(ghost_indices_vector.begin(),
+                                    ghost_indices_vector.end());
+
+
+    locally_active_dofs.subtract_set(locally_owned_dofs);
+
+    // 1) collect remote rows of sparse matrix
+    const auto locally_relevant_matrix_entries =
+      internal::extract_remote_rows<Number>(system_matrix,
+                                            sparsity_pattern,
+                                            locally_active_dofs,
+                                            internal::get_mpi_communicator(
+                                              system_matrix));
+
+
+    // 2) loop over all cells and "revert" assemly
+    blocks.clear();
+    blocks.resize(indices.size());
+
+    for (unsigned int c = 0; c < indices.size(); ++c)
+      {
+        if (indices[c].size() == 0)
+          continue;
```

Comment:
I think this needs to be documented - we support the case of empty indices, in which case the result is also empty (clearly).